### PR TITLE
Extend bus types and structure for Peripheral Manager

### PR DESCRIPTION
**Pull request by P-R-O-C-H-Y**
## Description of Change
This PR extends informations about the pin in Peripheral manager:

- Adds more specific bus types for peripherals pins.
- Adds `bus_num` option to specify bus number or unit.
- Adds `bus_channel` option to specify bus channel.
- Adds new custom option to set `extra_type` for the bus used on the pin.

All of those informations will be printed by `printPerimanInfo()` after end of `Setup()` if proper debug level is set.

+ Added new file `extra_attr.h` for easier use for some struct attributes.

**Output example:**

Before:
```
------------------------------------------
GPIO Info:
------------------------------------------
                  0 : ADC_ONESHOT
                  1 : UART_TX
                  3 : UART_RX
                  5 : GPIO
                 18 : SPI_MASTER
                 19 : SPI_MASTER
                 21 : I2C_MASTER
                 22 : I2C_MASTER
                 23 : SPI_MASTER
                 26 : UART_RX
                 27 : UART_TX
============ After Setup End =============

```
With changes in PR:
```
------------------------------------------
GPIO Info:
------------------------------------------
  GPIO : BUS_TYPE[bus/unit][chan]
  --------------------------------------  
     0 : ADC_ONESHOT[1][1]
     1 : UART_TX[0]
     3 : UART_RX[0]
     5 : SD_SS
    18 : SPI_MASTER_SCK[3]
    19 : SPI_MASTER_MISO[3]
    21 : I2C_MASTER_SDA[0]
    22 : I2C_MASTER_SCL[0]
    23 : SPI_MASTER_MOSI[3]
    26 : UART_RX[1]
    27 : UART_TX[1]
============ After Setup End =============
```

## Tests scenarios
Locally on ESP32 + CI.

## Related links

### CODE REVIEW

File: cores/esp32/esp32-hal-periman.h [hunk 1]
```diff
@@ -72,21 +109,33 @@ typedef bool (*peripheral_bus_deinit_cb_t)(void * bus);
...
 // Sets the peripheral destructor callback. Used to destroy bus when pin is assigned another function
 bool perimanSetBusDeinit(peripheral_bus_type_t type, peripheral_bus_deinit_cb_t cb);
 
 // Check if given pin is a valid GPIO number 
 bool perimanPinIsValid(uint8_t pin);
 
+// Sets extratype for non Init bus
+bool perimanSetPinBusExtraType(uint8_t pin, const char* extra_type);
+
+// Returns the extra tyoe of the bus for given pin if set. NULL otherwise
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-periman.h:136]: `type` typo

File: cores/esp32/esp32-hal-periman.h [hunk 2]
```diff
@@ -72,21 +109,33 @@ typedef bool (*peripheral_bus_deinit_cb_t)(void * bus);
...
+// Returns the bus channel of the bus for the given pin if set. -1 otherwise
+int8_t perimanGetPinBusChannel(uint8_t pin);
+
 // Sets the peripheral destructor callback. Used to destroy bus when pin is assigned another function
 bool perimanSetBusDeinit(peripheral_bus_type_t type, peripheral_bus_deinit_cb_t cb);
 
 // Check if given pin is a valid GPIO number 
 bool perimanPinIsValid(uint8_t pin);
 
+// Sets extratype for non Init bus
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-periman.h:133]: `extra type` and maybe try to explain it slightly better? :D

File: cores/esp32/esp32-hal-periman.h [hunk 3]
```diff
@@ -14,6 +14,8 @@ extern "C"
 #include <stdbool.h>
 #include <stddef.h>
 
+#define perimanDetachPin(p) perimanSetPinBus(p, ESP32_BUS_TYPE_INIT, NULL, -1, -1)
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-periman.h:17]: maybe `perimanClearPinBus` is a better name? it's more in line with `perimanSetPinBus`

**Comment by me-no-dev** [comment]
please post example sketch and output :)

File: cores/esp32/esp32-hal-periman.h (hunk 2 above)
Comment (P-R-O-C-H-Y) [reply cores/esp32/esp32-hal-periman.h]: Better now? :D

Comment (me-no-dev) [reply cores/esp32/esp32-hal-periman.h]: better :P

**Review by lucasssvaz — APPROVED** [review]
Tested it. Everything worked as expected.

File: cores/esp32/esp32-hal-periman.c [hunk 4]
```diff
@@ -8,9 +8,12 @@
 #include "esp32-hal-periman.h"
 #include "esp_bit_defs.h"
 
-typedef struct {
+typedef struct ATTR_PACKED {
 	peripheral_bus_type_t type;
+	const char* extra_type;
 	void * bus;
+	int8_t bus_num;
+	int8_t bus_channel;
```
Comment (SuGlider) [inline cores/esp32/esp32-hal-periman.c:16]: Shouldn't `extra_type` be changed to `description`? I couldn't see other usage for it...

Comment (me-no-dev) [reply cores/esp32/esp32-hal-periman.c:16]: it is an extra type. That is what is being used for. We added it for the cases where GPIO is being used for some specific task (like CS, RST, DC, etc.) and cases where different peripherals are used for some other specific task/driver. Hypothetic example would be implementing NeoPixel over RMT/Else and setting the extra type as `NEOPIXEL_OUT`

**Review by me-no-dev — APPROVED** [review]

**Review by SuGlider — APPROVED** [review]
Very Good Improvement @P-R-O-C-H-Y !

**Comment by SuGlider** [comment]
@me-no-dev - I think that we could already merge it.

**Pull request merged**